---
title: "${miseq_project_prefix}: ${analysis_type}"
output:
  html_notebook:
    df_print: paged
    number_sections: yes
    theme: lumen
    toc: yes
    code_folding: hide
---

<style>
  html {
    font-size: 16pt;
  }
  
  body {
    font-size: 16pt;
  }
  
  h1 {
    font-size: 2.2rem;
  }
  
  h2 {
    font-size: 2rem;
  }
  
  h3 {
    font-size: 1.8rem;
  }
  
  h4 {
    font-size: 1.4rem;
  }
  
</style>

```{r}
source('${project_metadata_file}')
analysis_type = "${analysis_type}"
clustering_level = "${clustering_level}"
tables_dir = "${tables_dir}"
analysis_title = tools::toTitleCase(gsub("_", " ", analysis_type))
```


# ${miseq_project_prefix} UNIFRAC Beta Diversity {.tabset}

## Clustering Level

**Clustering**
Data clustered at the `r toupper('${clustering_level}')` Level.

**Subgroups**
  ```{r, results='asis', echo=FALSE}
for (sg_filter in c(${subgroup_filters}))
{
  cat(sg_filter, '\n\n')
}
```

**Filtering**
* **Relative Abundance cutoff:** ${relative_abundance_cutoff}
* **Prevalence Cutoff:** ${prevalence_cutoff}
* **Min Count Cutoff:** ${min_count_cutoff}

```{r}
relative_abundance_cutoff = ${relative_abundance_cutoff}
prevalence_cutoff = ${prevalence_cutoff}
min_count_cutoff = ${min_count_cutoff}
```

## Setup

### Start Conda ENV
```{r}
startCondaEnv('betadiv', lib='~/R35')
```

### Load Libraries
```{r, echo=F}
library(tidyverse)
library(magrittr)
library(ggplot2)
library(ggbeeswarm)
library(openxlsx)
library(DT)
library(pheatmap)
library(kableExtra)
library(vegan)
library(phyloseq)
library(plotly)

### Custom libraries that can be loaded from GitHub
source('~/utilities/analysis-utilities/general_asv_data_analysis_utilities.R')
source('~/utilities/analysis-utilities/beta_diversity_utilities.R')
```

### Set Knitr Options
```{r}
${knitr_options}
```

### Load Data
```{r}
asv_table = 
  read.delim(
    '${asv_table_file}', 
    header=T, 
    sep='\t',
    stringsAsFactors=F
    )

taxonomy_table = 
  read.delim(
    '${taxonomy_table_file}', 
    header=T, 
    sep='\t',
    stringsAsFactors=F
    ) %>%
  mutate_all(as.character())

sample_data = 
  read.xlsx('${sample_data_file}') %>%
  mutate(
    CNV_Either_Eye = ifelse(CNV_Either_Eye == 1, '1', '0'),
    GA_No_CNV_Either_Eye = ifelse(GA_No_CNV_Either_Eye == 1, '1', '0')
  )
  
sample_data_columns = colnames(sample_data)
sample_names = sample_data$SampleName

load('${phylogenetic_tree_file}')


source(file.path(
  "${project_root}", 
  "${metadata_dir}", 
  'observational_variables.R'
  ))

```

### Extract Subgroups
```{r}
if (!is.null(c(${subgroup_filters})))
{
  subgroup_filter_string = paste0(${subgroup_filters}, collapse=" & ")
  print(sprintf("Subgroup Filter String: %s", subgroup_filter_string))
  subgroup_filter_expression = str2expression(subgroup_filter_string)
  sample_data = sample_data %>% filter(eval(subgroup_filter_expression))
  sample_names = sample_data$SampleName

  asv_table = asv_table %>% select('ASVs', sample_names)
}

```

### Make Phyloseq Object
```{r}
ps = phyloseq(
  otu_table(asv_table %>%
              remove_rownames() %>%
              column_to_rownames('ASVs') %>%
              as.matrix(),
    taxa_are_rows=T
  ),
  tax_table(taxonomy_table %>%
              remove_rownames() %>%
              column_to_rownames('ASVs') %>%
              as.matrix()
  ),
  sample_data(sample_data %>%
                remove_rownames() %>%
                column_to_rownames('SampleName')
              ),
  phy_tree(fit$tree)
)
```

### Variables of interest
#### Observational Variables
```{r}
case = ${case}
key_covariates = ${key_covariates}
treatments = ${treatments}
variables = c(case, key_covariates, treatments)
# categorical_variables = ${categorical_variables}
categorical_variables = c(case, key_covariates, treatments) %>% setdiff('Age')
```

### Variables of interest
#### Observational Variables
```{r}
case = ${case} %>% setdiff('')
genotypes = ${genotypes} %>% setdiff('')
phenotypes = ${phenotypes} %>% setdiff('')
key_covariates = ${key_covariates} %>% setdiff('')
treatments = ${treatments} %>% setdiff('')
categorical_variables = ${categorical_variables} %>% setdiff('')

variables = c(case, key_covariates, treatments, phenotypes, genotypes)
amd_only_variables = c(key_covariates, treatments, phenotypes, genotypes)
```

* **Case:** `r case`
* **Genotype variables:** `r genotypes`
* **Phenotype variables:** `r phenotypes`
* **Treatment variables:** `r treatments`
* **Key covariates:**  `r key_covariates`


#### Remove Age
```{r}
variables %<>% setdiff('Age')
```

* **Treatment variables:** ${treatments}
* **Key covariates:**  ${key_covariates}
* **Variables Used:** `r variables`

#### Axes Sets
```{r}
axes_sets = list(
  c(1,2),
  c(2,3),
  c(1,3)
)
```


## UNWEIGHTED
### All
#### Calculate Ordination
```{r}
ps_ord = ordinate(ps, 'PCoA', 'UNIFRAC', weighted=F)
```

#### Scree Plot
```{r}
plot_scree(ps_ord, "Unifrac, Unweighted")
```

#### 2D
```{r}
for (var in categorical_variables)
{
  var_data = observational_variables[[var]]
  print(var_data)
  
  remove_idx = 
    sample_data %>%
    mutate(
      remove_idx = (!!as.name(var) %in% c(var_data$case, var_data$control))
      ) %>%
    pull(remove_idx)
  
  this_ps = prune_samples(remove_idx, ps)
  
  ps_ord = ordinate(this_ps, 'PCoA', 'UNIFRAC', weighted=F)
  
  for (axes in axes_sets)
  {
    print(axes)
    
    plotUnifracOrd(
      this_ps, 
      ps_ord, 
      ord_type = 'PCoA',
      dist_type='UNIFRAC',
      weighted=F, 
      axes=axes,
      var_data = observational_variables[[var]],
      color=var,
      plot_type='samples',
      subtitle="${analysis_subtitle}"
    ) %>% print()
  }
  
}
```

#### 3D
```{r}
princomp_master_table = 
  ps_ord$vectors %>%
  data.frame() %>%
  rownames_to_column('SampleName') %>%
  left_join(sample_data, by='SampleName')
  
plotOrd3D(
  princomp_master_table=princomp_master_table,
  axes=paste0('Axis.', c(1,2,3)),
  id_col="SampleName",
  color="CaseString",
  title_extra="Unweighted Unifrac"
)
``` 

### Age < 90
#### Calculate Ordination
```{r}
ps_age_lt_90 = ps %>% subset_samples(Age < 90)
ps_ord = ordinate(ps_age_lt_90, 'PCoA', 'UNIFRAC', weighted=F)
```

#### Scree Plot
```{r}
plot_scree(ps_ord, "Unifrac, Unweighted, Age<90")
```

#### 2D
```{r}
for (var in categorical_variables)
{
  var_data = observational_variables[[var]]
  print(var_data)
  
  remove_idx = 
    sample_data %>%
    filter(Age < 90) %>%
    mutate(
      remove_idx = (!!as.name(var) %in% c(var_data$case, var_data$control))
      ) %>%
    pull(remove_idx)
  
  this_ps = prune_samples(remove_idx, ps_age_lt_90)
  
  ps_ord = ordinate(this_ps, 'PCoA', 'UNIFRAC', weighted=F)
  
  for (axes in axes_sets)
  {
    print(axes)
    
    plotUnifracOrd(
      this_ps, 
      ps_ord, 
      ord_type = 'PCoA',
      dist_type='UNIFRAC',
      weighted=F, 
      axes=axes,
      var_data = observational_variables[[var]],
      color=var,
      plot_type='samples',
      subtitle="${analysis_subtitle}"
    ) %>% print()
  }
  
}
```


#### 3D
```{r}
princomp_master_table = 
  ps_ord$vectors %>%
  data.frame() %>%
  rownames_to_column('SampleName') %>%
  left_join(sample_data, by='SampleName')
  
plotOrd3D(
  princomp_master_table=princomp_master_table,
  axes=paste0('Axis.', c(1,2,3)),
  id_col="SampleName",
  color="CaseString",
  title_extra="Unweighted Unifrac, Age < 90"
)
```

## WEIGHTED
### All
#### Calculate Ordination
```{r}
ps_ord = ordinate(ps, 'PCoA', 'UNIFRAC', weighted=T)
```

#### Scree PLot
```{r}
plot_scree(ps_ord, "Unifrac Weighted")
```

#### 2D
```{r}
for (var in categorical_variables)
{
  var_data = observational_variables[[var]]
  print(var_data)
  
  remove_idx = 
    sample_data %>%
    mutate(
      remove_idx = (!!as.name(var) %in% c(var_data$case, var_data$control))
      ) %>%
    pull(remove_idx)
  
  this_ps = prune_samples(remove_idx, ps)
  
  ps_ord = ordinate(this_ps, 'PCoA', 'UNIFRAC', weighted=T)
  
  for (axes in axes_sets)
  {
    print(axes)
    
    plotUnifracOrd(
      this_ps, 
      ps_ord, 
      ord_type = 'PCoA',
      dist_type='UNIFRAC',
      weighted=T, 
      axes=axes,
      var_data = observational_variables[[var]],
      color=var,
      plot_type='samples',
      subtitle="${analysis_subtitle}"
    ) %>% print()
  }
  
}
```

#### 3D
```{r}
princomp_master_table = 
  ps_ord$vectors %>%
  data.frame() %>%
  rownames_to_column('SampleName') %>%
  left_join(sample_data, by='SampleName')
  
plotOrd3D(
  princomp_master_table=princomp_master_table,
  axes=paste0('Axis.', c(1,2,3)),
  id_col="SampleName",
  color="CaseString",
  title_extra="Weighted Unifrac"
)
```

### Age < 90
#### Calculate Ordination
```{r}
ps_age_lt_90 = ps %>% subset_samples(Age < 90)
ps_ord = ordinate(ps_age_lt_90, 'PCoA', 'UNIFRAC', weighted=T)
```

#### Scree PLot
```{r}
plot_scree(ps_ord, "Unifrac Weighted, Age < 90")
```

#### 2D
```{r}
for (var in categorical_variables)
{
  var_data = observational_variables[[var]]
  print(var_data)
  
  remove_idx = 
    sample_data %>%
    filter(Age < 90) %>%
    mutate(
      remove_idx = (!!as.name(var) %in% c(var_data$case, var_data$control))
      ) %>%
    pull(remove_idx)
  
  this_ps = prune_samples(remove_idx, ps_age_lt_90)
  
  ps_ord = ordinate(this_ps, 'PCoA', 'UNIFRAC', weighted=F)
  
  for (axes in axes_sets)
  {
    print(axes)
    
    plotUnifracOrd(
      this_ps, 
      ps_ord, 
      ord_type = 'PCoA',
      dist_type='UNIFRAC',
      weighted=F, 
      axes=axes,
      var_data = observational_variables[[var]],
      color=var,
      plot_type='samples',
      subtitle="${analysis_subtitle}"
    ) %>% print()
  }
  
}

```

#### 3D
```{r}
princomp_master_table = 
  ps_ord$vectors %>%
  data.frame() %>%
  rownames_to_column('SampleName') %>%
  left_join(sample_data, by='SampleName')
  
plotOrd3D(
  princomp_master_table=princomp_master_table,
  axes=paste0('Axis.', c(1,2,3)),
  id_col="SampleName",
  color="CaseString",
  title_extra="Weighted Unifrac, Age<90"
)
```
