---
title: "MiSeq-${miseq_run_number}"
output:
  html_notebook:
    df_print: paged
    number_sections: yes
    theme: lumen
    toc: yes
    code_folding: hide
---

<style>
  html {
    font-size: 16pt;
  }
  
  body {
    font-size: 16pt;
  }
  
  h1 {
    font-size: 2.2rem;
  }
  
  h2 {
    font-size: 2rem;
  }
  
  h3 {
    font-size: 1.8rem;
  }
  
  h4 {
    font-size: 1.4rem;
  }
  
  .dataTable {
    font-size: 9pt;
  }
  
</style>

```{r}
source('${project_metadata_file}')
analysis_type = "${analysis_type}"
clustering_level = "${clustering_level}"
tables_dir = "${tables_dir}"
use_allbac = ${use_allbac}
analysis_title = tools::toTitleCase(gsub("_", " ", analysis_type))
```

# MiSeq-${miseq_run_number} DESeq2 Analysis {.tabset}

## Clustering
Data clustered at the `r toupper("${clustering_level}")` level.

## Filtering
* **Relative Abundance Cutoff:** ${relative_abundance_cutoff}
* **Prevalence Cutoff:** ${prevalence_cutoff}
* **Min Count Cutoff:** ${min_count_cutoff}


## Setup 

### [Close]

### Start Conda ENV
```{r}
startCondaEnv('deseq', lib='~/R35')
```

### Load Libraries
```{r}
library(rmarkdown)
library(tidyverse)
library(magrittr)
library(ggplot2)
library(openxlsx)
library(DT)
library(kableExtra)
library(DESeq2)

### Custom libraries that can be loaded from GitHub
source('~/utilities/analysis-utilities/general_asv_data_analysis_utilities.R')
source('~/utilities/analysis-utilities/deseq_utilities.R')
source('~/utilities/amd_templates/setup/amd_project_utilities.R')
```

### Set Knitr Options
```{r}
${knitr_options}
```

### Load Data
```{r}
asv_table = 
  read.delim(
    '${asv_table_file}', 
    header=T, 
    sep='\t',
    stringsAsFactors=F
    )

taxonomy_table = 
  read.delim(
    '${taxonomy_table_file}', 
    header=T, 
    sep='\t',
    stringsAsFactors=F
    ) %>%
  mutate_all(as.character())

sample_data = read.xlsx('${sample_data_file}')

if ("IGA" %in% colnames(sample_data) & use_allbac)
{
  sample_data = 
    sample_data %>%
    filter(IGA=='AllBac')

  sample_names = sample_data$SampleName
  
  asv_table = 
    asv_table %>%
    select('ASVs', ends_with('AllBac'))
}

sample_data_columns = colnames(sample_data)
sample_names = sample_data$SampleName

source(file.path(
  "${project_root}", 
  "${metadata_dir}", 
  'observational_variables.R'
  ))

load(file.path(dada2_table_dir, '${phylogenetic_tree_file}'))
```


### Set Metadata Factor Levels
```{r}
sample_data = setFactorsLevels(sample_data, observational_variables)
```

### Variables of interest
#### Observational Variables
```{r}
genotypes = ${genotypes}
phenotypes = ${phenotypes}
key_covariates = ${key_covariates}
treatments = ${treatments}
amd_only_variables = c(key_covariates, treatments, phenotypes, genotypes) %>% unique()
```

#### Remove Age
```{r}
key_covariates %<>% setdiff('Age')
amd_only_variables %<>% setdiff('Age')
```

* **Genotype variables:** `r genotypes`
* **Phenotype variables:** `r phenotypes`
* **Treatment variables:** `r treatments`
* **Other covariates:**  `r key_covariates`

### Set Filtering
```{r}
relative_abundance_cutoff=${relative_abundance_cutoff}
prevalence_cutoff=${prevalence_cutoff}
```

## Variables Independantly
### Create Workbook
```{r}
deseq_results_wb = createWorkbook()
```

### Run and add worksheets
```{r}

lowest_rank = 'Genus'

for (varname in amd_only_variables)
{
  
  var = observational_variables[[varname]]
  
  taxa_counts = getFilteredTaxaCountsDev(
    asv_table=asv_table,
    taxonomy_table=taxonomy_table,
    sample_data=sample_data,
    cluster_by=lowest_rank,
    relative_abundance_cutoff=relative_abundance_cutoff,
    prevalence_cutoff=prevalence_cutoff,
    min_count_cutoff=0,
    filter_by='Taxa',
    clean_taxa=T,  ### remove NAs in lowest rank
    n_max_by_mean=F,
    id_col="SampleName", ### metadata column that containes the unique sample IDs
    add_glommed_names=T,
    normalize=F
  )
  
  if (var$case != "" && var$control != "")
  {
    var_sample_data =
      sample_data %>% 
      filter(!!as.name(varname) %in% c(var$case, var$control))
  } else
  {
    var_sample_data = sample_data
  }
    
  results_df = runDeseqFromTables_dev(
    asv_table,
    taxonomy_table,
    var_sample_data,
    varname,
    lowest_rank=lowest_rank,
    relative_abundance_cutoff=relative_abundance_cutoff,
    prevalence_cutoff=prevalence_cutoff,
    include_covariates = 'All'
    )
  
    match_string = paste0(varname, '.*\\.padj')
    padj_col = grep(match_string, colnames(results_df), value=T)
  
  results_df =
    results_df %>%
    select(Taxon, !!padj_col, starts_with(varname), everything()) %>%
    arrange(!!as.name(padj_col))
  
  addWorksheet(
    wb=deseq_results_wb, 
    sheetName=varname
    )
  
  openxlsx::writeData(
    wb=deseq_results_wb, 
    sheet=varname, 
    x=results_df
    )
  
  setColWidths(
    wb=deseq_results_wb, 
    sheet=varname, 
    cols=1, 
    widths = "auto"
  )
  
  freezePane(
    deseq_results_wb, 
    sheet=varname, 
    firstCol=T,
    firstRow=T
    )
}
```

### Save Workbook
```{r}
filename = makeDataFileName(
  'amd_only_individual_variables.xlsx',
  "${tables_dir}",
  "${analysis_type}",
  "${miseq_project_prefix}",
  "rank"
)

print(filename)

saveWorkbook(deseq_results_wb, file=filename, overwrite=T)
```

## Genotypes, Phenotypes, and AREDS with Controls
### Create Workbook
```{r}
deseq_results_wb = createWorkbook()
```

### Run and add worksheets
```{r}
lowest_rank = 'Genus'

for (varname in c(genotypes, phenotypes, treatments))
{
  
  vars = observational_variables[c(varname, key_covariates)]
  
  risk = vars[[varname]]$case
  non_risk = vars[[varname]]$control
  
  taxa_counts = getFilteredTaxaCountsDev(
    asv_table=asv_table,
    taxonomy_table=taxonomy_table,
    sample_data=sample_data,
    cluster_by=lowest_rank,
    relative_abundance_cutoff=relative_abundance_cutoff,
    prevalence_cutoff=prevalence_cutoff,
    min_count_cutoff=0,
    filter_by='Taxa',
    clean_taxa=T,  ### remove NAs in lowest rank
    n_max_by_mean=F,
    id_col="SampleName", ### metadata column that containes the unique sample IDs
    add_glommed_names=T,
    normalize=F
  )
    
  results_df = runDeseqFromTables_dev(
    asv_table,
    taxonomy_table,
    sample_data %>% filter(!!as.name(varname) %in% c(risk, non_risk)),
    vars %>% names(),
    lowest_rank=lowest_rank,
    relative_abundance_cutoff=relative_abundance_cutoff,
    prevalence_cutoff=prevalence_cutoff,
    include_covariates = 'All'
    )
  
  match_string = paste0(varname, '.*\\.padj')
  padj_col = grep(match_string, colnames(results_df), value=T)
  
  results_df =
    results_df %>%
    select(Taxon, !!padj_col, starts_with(varname), everything()) %>%
    arrange(!!as.name(padj_col))
    
  addWorksheet(
    wb=deseq_results_wb, 
    sheetName=varname
    )
  
  openxlsx::writeData(
    wb=deseq_results_wb, 
    sheet=varname, 
    x=results_df
    )
  
  setColWidths(
    wb=deseq_results_wb, 
    sheet=varname, 
    cols=1, 
    widths = "auto"
  )
  
  freezePane(
    deseq_results_wb, 
    sheet=varname, 
    firstCol=T,
    firstRow=T
    )

}
```

### Save Workbook
```{r}
filename = makeDataFileName(
  'amd_only_variables_with_controls.xlsx',
  "${tables_dir}",
  "${analysis_type}",
  "${miseq_project_prefix}",
  "rank"
)

print(filename)

saveWorkbook(deseq_results_wb, file=filename, overwrite=T)
```
